File: Projects/Offensive/Firmware reverse-engineering/firmware_analyzer.py
#!/usr/bin/env python3
"""
Firmware Reverse Engineering Toolkit
Purpose: Analyze embedded firmware for security vulnerabilities
Use: Security assessment, vulnerability research, IoT device testing
"""

import os
import struct
import hashlib
import binascii
import argparse
import re
from pathlib import Path
from typing import Dict, List, Optional

class FirmwareAnalyzer:
    def __init__(self, firmware_path):
        self.firmware_path = firmware_path
        self.firmware_data = None
        self.analysis_results = {}
        self.vulnerabilities = []
        
    def load_firmware(self):
        """Load firmware file into memory"""
        try:
            with open(self.firmware_path, 'rb') as f:
                self.firmware_data = f.read()
            print(f"Loaded firmware: {len(self.firmware_data)} bytes")
            return True
        except Exception as e:
            print(f"Error loading firmware: {e}")
            return False
    
    def identify_file_type(self):
        """Identify firmware file type using magic bytes"""
        magic_bytes = {
            b'\x7fELF': 'ELF Executable',
            b'\xce\xfa\xed\xfe': 'Mach-O Binary',
            b'MZ': 'Windows PE',
            b'\x27\x05\x19\x56': 'uImage',
            b'\x55\xaa': 'MBR Bootloader',
            b'UBI#': 'UBI Image',
            b'TRX': 'TRX Firmware',
            b'HSQS': 'SquashFS',
        }
        
        for magic, file_type in magic_bytes.items():
            if self.firmware_data.startswith(magic):
                return file_type
        
        # Check for common compressed formats
        if self.firmware_data.startswith(b'\x1f\x8b'):
            return 'GZIP Compressed'
        elif self.firmware_data.startswith(b'\x42\x5a\x68'):
            return 'BZIP2 Compressed'
        elif self.firmware_data.startswith(b'\xfd7zXZ'):
            return 'XZ Compressed'
        
        return 'Unknown'
    
    def extract_strings(self, min_length=4):
        """Extract ASCII strings from firmware"""
        strings = []
        current_string = ""
        
        for byte in self.firmware_data:
            if 32 <= byte <= 126:  # Printable ASCII range
                current_string += chr(byte)
            else:
                if len(current_string) >= min_length:
                    strings.append(current_string)
                current_string = ""
        
        # Don't forget the last string
        if len(current_string) >= min_length:
            strings.append(current_string)
        
        return strings
    
    def analyze_encryption(self):
        """Check for encryption and weak cryptography"""
        crypto_indicators = {
            'MD5': hashlib.md5(b'test').digest()[:4],
            'SHA1': hashlib.sha1(b'test').digest()[:4],
            'AES': b'\x63\x20\x63\x68',  # Common AES constants
            'DES': b'\x0e\x32\xfc\xf7',
            'RSA': b'\x30\x82',  # PKCS#8 header
        }
        
        crypto_found = []
        for algo, pattern in crypto_indicators.items():
            if pattern in self.firmware_data:
                crypto_found.append(algo)
        
        return crypto_found
    
    def search_hardcoded_secrets(self):
        """Search for hardcoded credentials and secrets"""
        secrets_found = []
        
        # Common password patterns
        password_patterns = [
            rb'password[\s=:]+["\']([^"\']+)["\']',
            rb'passwd[\s=:]+["\']([^"\']+)["\']',
            rb'pwd[\s=:]+["\']([^"\']+)["\']',
            rb'admin:([^:\n]+)',
            rb'root:([^:\n]+)',
        ]
        
        strings = self.extract_strings()
        for string in strings:
            # Check for API keys
            if re.match(r'[A-Za-z0-9]{32}', string):
                secrets_found.append(f"Possible API Key: {string}")
            
            # Check for default passwords
            common_passwords = ['admin', 'password', '1234', 'default', 'root']
            if string.lower() in common_passwords:
                secrets_found.append(f"Default Password: {string}")
        
        return secrets_found
    
    def analyze_network_services(self):
        """Identify potential network services and ports"""
        network_indicators = []
        
        # Common port numbers in binary
        common_ports = [80, 443, 22, 23, 21, 25, 53, 8080, 8443]
        for port in common_ports:
            # Search for port in big and little endian
            port_be = struct.pack('>H', port)
            port_le = struct.pack('<H', port)
            
            if port_be in self.firmware_data or port_le in self.firmware_data:
                network_indicators.append(f"Port {port} reference found")
        
        # Search for common network function names
        network_functions = ['socket', 'bind', 'listen', 'accept', 'connect',
                           'send', 'recv', 'httpd', 'telnetd', 'sshd']
        
        for func in network_functions:
            if func.encode() in self.firmware_data:
                network_indicators.append(f"Network function: {func}")
        
        return network_indicators
    
    def check_binary_protections(self):
        """Check for common binary security features"""
        protections = {}
        
        # This is a simplified check - real implementation would use binwalk or similar
        strings = self.extract_strings()
        
        # Check for stack protection
        stack_protection_indicators = ['stack_chk_fail', '__stack_chk_guard']
        if any(indicator in strings for indicator in stack_protection_indicators):
            protections['Stack Canary'] = 'Present'
        else:
            protections['Stack Canary'] = 'Absent'
        
        # Check for ASLR/PIE (simplified)
        if b'/proc/self/exe' in self.firmware_data:
            protections['ASLR/PIE'] = 'Possible'
        else:
            protections['ASLR/PIE'] = 'Unknown'
        
        return protections
    
    def search_vulnerable_functions(self):
        """Search for potentially dangerous functions"""
        dangerous_functions = {
            'strcpy': 'Use strncpy instead',
            'gets': 'Extremely dangerous - use fgets',
            'sprintf': 'Use snprintf to prevent buffer overflow',
            'strcat': 'Use strncat with proper bounds checking',
            'system': 'Command injection risk',
            'popen': 'Command injection risk',
        }
        
        found_functions = []
        for func, risk in dangerous_functions.items():
            if func.encode() in self.firmware_data:
                found_functions.append(f"{func}: {risk}")
        
        return found_functions
    
    def generate_hex_dump(self, offset=0, length=512):
        """Generate hex dump of firmware section"""
        if offset + length > len(self.firmware_data):
            length = len(self.firmware_data) - offset
        
        data_slice = self.firmware_data[offset:offset + length]
        hex_dump = binascii.hexlify(data_slice).decode('ascii')
        
        # Format with 16 bytes per line
        formatted = []
        for i in range(0, len(hex_dump), 32):
            line = hex_dump[i:i+32]
            ascii_repr = ''.join(chr(int(line[j:j+2], 16)) 
                               if 32 <= int(line[j:j+2], 16) <= 126 else '.' 
                               for j in range(0, len(line), 2))
            formatted.append(f"{offset + i//2:08x}: {line}  {ascii_repr}")
        
        return '\n'.join(formatted)
    
    def comprehensive_analysis(self):
        """Perform comprehensive firmware analysis"""
        if not self.load_firmware():
            return False
        
        print("Starting firmware analysis...")
        print("=" * 50)
        
        # File type identification
        file_type = self.identify_file_type()
        print(f"File Type: {file_type}")
        
        # Cryptographic analysis
        crypto = self.analyze_encryption()
        if crypto:
            print(f"Crypto Found: {', '.join(crypto)}")
        
        # Security protections
        protections = self.check_binary_protections()
        print("Binary Protections:")
        for protection, status in protections.items():
            print(f"  {protection}: {status}")
        
        # Network services
        network_services = self.analyze_network_services()
        if network_services:
            print("Network Services Found:")
            for service in network_services[:5]:  # Show first 5
                print(f"  {service}")
        
        # Vulnerable functions
        dangerous_funcs = self.search_vulnerable_functions()
        if dangerous_funcs:
            print("Potentially Dangerous Functions:")
            for func in dangerous_funcs:
                print(f"  {func}")
        
        # Hardcoded secrets
        secrets = self.search_hardcoded_secrets()
        if secrets:
            print("Hardcoded Secrets Found:")
            for secret in secrets[:5]:  # Show first 5
                print(f"  {secret}")
        
        # Sample hex dump
        print("\nSample Hex Dump (first 512 bytes):")
        print(self.generate_hex_dump())
        
        return True

class FirmwareExtractor:
    """Advanced firmware extraction and analysis"""
    
    def __init__(self):
        self.compressions = {
            b'\x1f\x8b': 'gzip',
            b'\x42\x5a\x68': 'bzip2', 
            b'\xfd7zXZ': 'xz',
            b'\x50\x4b\x03\x04': 'zip',
        }
    
    def extract_filesystem(self, firmware_data, output_dir):
        """Attempt to extract embedded filesystem"""
        try:
            import binwalk
            # This would use binwalk API for real extraction
            print("Filesystem extraction would require binwalk integration")
            return False
        except ImportError:
            print("Binwalk not available for advanced extraction")
            return False
    
    def analyze_elf_sections(self, firmware_path):
        """Analyze ELF binary sections"""
        try:
            import elftools
            from elftools.elf.elffile import ELFFile
            
            with open(firmware_path, 'rb') as f:
                elf = ELFFile(f)
                
                print("ELF Sections:")
                for section in elf.iter_sections():
                    print(f"  {section.name}: {section['sh_size']} bytes")
                    
            return True
        except ImportError:
            print("pyelftools required for ELF analysis")
            return False

def main():
    parser = argparse.ArgumentParser(description='Firmware Reverse Engineering Tool')
    parser.add_argument('firmware', help='Path to firmware file')
    parser.add_argument('--extract', action='store_true', help='Attempt filesystem extraction')
    parser.add_argument('--strings', action='store_true', help='Extract strings only')
    parser.add_argument('--crypto', action='store_true', help='Cryptographic analysis only')
    
    args = parser.parse_args()
    
    if not os.path.exists(args.firmware):
        print(f"Error: Firmware file {args.firmware} not found")
        return
    
    analyzer = FirmwareAnalyzer(args.firmware)
    
    if args.strings:
        analyzer.load_firmware()
        strings = analyzer.extract_strings()
        print(f"Found {len(strings)} strings:")
        for string in strings[:50]:  # Show first 50
            print(f"  {string}")
    elif args.crypto:
        analyzer.load_firmware()
        crypto = analyzer.analyze_encryption()
        print("Cryptographic analysis:")
        print(f"  Algorithms found: {crypto}")
    else:
        analyzer.comprehensive_analysis()

if __name__ == "__main__":
    main()

File: Projects/Offensive/Firmware reverse-engineering/firmware_analyzer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/stat.h>

#define MAX_STRING_LENGTH 256
#define BUFFER_SIZE 4096
#define MAX_STRINGS 10000

typedef struct {
    unsigned char *data;
    size_t size;
    char *filename;
} Firmware;

typedef struct {
    char **strings;
    int count;
    int capacity;
} StringList;

typedef struct {
    char *vulnerability;
    char *description;
    int severity; // 1-10 scale
} Vulnerability;

// Basic firmware analysis functions
Firmware* load_firmware(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        printf("Error: Cannot open firmware file %s\n", filename);
        return NULL;
    }
    
    // Get file size
    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);
    
    if (file_size <= 0) {
        printf("Error: Invalid file size\n");
        fclose(file);
        return NULL;
    }
    
    // Allocate memory for firmware
    Firmware *fw = malloc(sizeof(Firmware));
    if (!fw) {
        printf("Error: Memory allocation failed\n");
        fclose(file);
        return NULL;
    }
    
    fw->data = malloc(file_size);
    if (!fw->data) {
        printf("Error: Memory allocation failed\n");
        free(fw);
        fclose(file);
        return NULL;
    }
    
    // Read file data
    size_t bytes_read = fread(fw->data, 1, file_size, file);
    fclose(file);
    
    if (bytes_read != file_size) {
        printf("Error: File read incomplete\n");
        free(fw->data);
        free(fw);
        return NULL;
    }
    
    fw->size = file_size;
    fw->filename = strdup(filename);
    
    printf("Loaded firmware: %s (%zu bytes)\n", filename, file_size);
    return fw;
}

void free_firmware(Firmware *fw) {
    if (fw) {
        free(fw->data);
        free(fw->filename);
        free(fw);
    }
}

// String extraction functions
int is_printable_ascii(unsigned char c) {
    return (c >= 32 && c <= 126);
}

StringList* extract_strings(Firmware *fw, int min_length) {
    StringList *list = malloc(sizeof(StringList));
    if (!list) return NULL;
    
    list->capacity = MAX_STRINGS;
    list->count = 0;
    list->strings = malloc(sizeof(char*) * list->capacity);
    
    char current_string[MAX_STRING_LENGTH];
    int string_index = 0;
    
    for (size_t i = 0; i < fw->size; i++) {
        unsigned char c = fw->data[i];
        
        if (is_printable_ascii(c)) {
            if (string_index < MAX_STRING_LENGTH - 1) {
                current_string[string_index++] = c;
            }
        } else {
            if (string_index >= min_length) {
                current_string[string_index] = '\0';
                
                // Add to list
                if (list->count < list->capacity) {
                    list->strings[list->count] = strdup(current_string);
                    list->count++;
                }
            }
            string_index = 0;
        }
    }
    
    // Handle last string if file doesn't end with non-printable
    if (string_index >= min_length) {
        current_string[string_index] = '\0';
        if (list->count < list->capacity) {
            list->strings[list->count] = strdup(current_string);
            list->count++;
        }
    }
    
    printf("Extracted %d strings\n", list->count);
    return list;
}

void free_string_list(StringList *list) {
    if (list) {
        for (int i = 0; i < list->count; i++) {
            free(list->strings[i]);
        }
        free(list->strings);
        free(list);
    }
}

// Pattern matching for vulnerability detection
int search_pattern(Firmware *fw, const unsigned char *pattern, size_t pattern_len) {
    int matches = 0;
    
    for (size_t i = 0; i <= fw->size - pattern_len; i++) {
        if (memcmp(&fw->data[i], pattern, pattern_len) == 0) {
            matches++;
        }
    }
    
    return matches;
}

// Check for common dangerous functions
void check_dangerous_functions(Firmware *fw) {
    struct {
        const char *name;
        const char *risk;
    } dangerous_funcs[] = {
        {"strcpy", "Buffer overflow risk"},
        {"gets", "Extremely dangerous - no bounds checking"},
        {"sprintf", "Buffer overflow risk"},
        {"strcat", "Buffer overflow risk"},
        {"system", "Command injection risk"},
        {"popen", "Command injection risk"},
        {NULL, NULL}
    };
    
    printf("\nDangerous Function Analysis:\n");
    
    for (int i = 0; dangerous_funcs[i].name != NULL; i++) {
        int matches = search_pattern(fw, (unsigned char*)dangerous_funcs[i].name, 
                                   strlen(dangerous_funcs[i].name));
        if (matches > 0) {
            printf("  [WARNING] %s: %s (%d occurrences)\n", 
                   dangerous_funcs[i].name, dangerous_funcs[i].risk, matches);
        }
    }
}

// Check for hardcoded credentials
void check_hardcoded_credentials(StringList *strings) {
    const char *common_passwords[] = {
        "admin", "password", "1234", "default", "root", 
        "pass", "pwd", "123456", "guest", NULL
    };
    
    printf("\nHardcoded Credentials Check:\n");
    
    for (int i = 0; common_passwords[i] != NULL; i++) {
        for (int j = 0; j < strings->count; j++) {
            if (strcmp(strings->strings[j], common_passwords[i]) == 0) {
                printf("  [WARNING] Found default password: %s\n", common_passwords[i]);
                break;
            }
        }
    }
}

// File type identification
void identify_file_type(Firmware *fw) {
    struct {
        const unsigned char *magic;
        size_t length;
        const char *description;
    } magic_numbers[] = {
        {(unsigned char*)"\x7f""ELF", 4, "ELF Executable"},
        {(unsigned char*)"MZ", 2, "Windows PE Executable"},
        {(unsigned char*)"\x1f\x8b", 2, "GZIP Compressed"},
        {(unsigned char*)"\x42\x5a\x68", 3, "BZIP2 Compressed"},
        {(unsigned char*)"\xfd""7zXZ", 6, "XZ Compressed"},
        {(unsigned char*)"\x55\xaa", 2, "MBR Bootloader"},
        {NULL, 0, NULL}
    };
    
    printf("\nFile Type Identification:\n");
    
    for (int i = 0; magic_numbers[i].magic != NULL; i++) {
        if (fw->size >= magic_numbers[i].length) {
            if (memcmp(fw->data, magic_numbers[i].magic, magic_numbers[i].length) == 0) {
                printf("  File Type: %s\n", magic_numbers[i].description);
                return;
            }
        }
    }
    
    printf("  File Type: Unknown\n");
}

// Generate hex dump
void generate_hex_dump(Firmware *fw, size_t offset, size_t length) {
    if (offset >= fw->size) return;
    
    if (offset + length > fw->size) {
        length = fw->size - offset;
    }
    
    printf("\nHex Dump (offset: 0x%zx, length: %zu):\n", offset, length);
    printf("Offset    Hexadecimal                           ASCII\n");
    printf("--------  ------------------------------------  ----------------\n");
    
    for (size_t i = 0; i < length; i += 16) {
        printf("%08zx  ", offset + i);
        
        // Hexadecimal portion
        for (size_t j = 0; j < 16; j++) {
            if (i + j < length) {
                printf("%02x ", fw->data[offset + i + j]);
            } else {
                printf("   ");
            }
            
            if (j == 7) printf(" ");
        }
        
        printf(" ");
        
        // ASCII portion
        for (size_t j = 0; j < 16; j++) {
            if (i + j < length) {
                unsigned char c = fw->data[offset + i + j];
                printf("%c", is_printable_ascii(c) ? c : '.');
            } else {
                printf(" ");
            }
        }
        
        printf("\n");
    }
}

// Main analysis function
void analyze_firmware(Firmware *fw) {
    printf("\n=== Firmware Security Analysis ===\n");
    
    // File type identification
    identify_file_type(fw);
    
    // Extract and analyze strings
    StringList *strings = extract_strings(fw, 4);
    
    // Check for dangerous functions
    check_dangerous_functions(fw);
    
    // Check for hardcoded credentials
    check_hardcoded_credentials(strings);
    
    // Generate sample hex dump
    generate_hex_dump(fw, 0, 256);
    
    // Show some interesting strings
    printf("\nInteresting Strings Found (first 20):\n");
    int shown = 0;
    for (int i = 0; i < strings->count && shown < 20; i++) {
        // Show strings that might be interesting
        if (strlen(strings->strings[i]) > 8) {
            printf("  %s\n", strings->strings[i]);
            shown++;
        }
    }
    
    free_string_list(strings);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <firmware_file>\n", argv[0]);
        return 1;
    }
    
    printf("Firmware Reverse Engineering Tool\n");
    printf("==================================\n");
    
    // Load firmware
    Firmware *fw = load_firmware(argv[1]);
    if (!fw) {
        return 1;
    }
    
    // Perform analysis
    analyze_firmware(fw);
    
    // Cleanup
    free_firmware(fw);
    
    return 0;
}
